src/governance.cpp

```diff
@@ -918,14 +919,18 @@ bool CGovernanceManager::MasternodeRateCheck(const CGovernanceObject& govobj, up
 
     switch(eUpdateLast) {
     case UPDATE_TRUE:
+        // we should update the buffer only if object has accepted
+        if (!fRateOK)
+            LogPrintf("CGovernanceManager::MasternodeRateCheck -- WARNING !!! Rate too high but eUpdateLast = UPDATE_TRUE : object hash = %s, masternode vin = %s, object timestamp = %d, rate = %f, max rate = %f\n",
+                        strHash, vin.prevout.ToStringShort(), nTimestamp, dRate, dMaxRate);
         pBuffer->AddTimestamp(nTimestamp);
         it->second.fStatusOK = fRateOK;
-        break;
+        return true;
     case UPDATE_FAIL_ONLY:
-        if(!fRateOK) {
-            pBuffer->AddTimestamp(nTimestamp);
+        if(!fRateOK)
```

Comment:
Actually, we pass `UPDATE_TRUE` only when the object has been accepted. Would it better to have a separate method for updating the buffer? I have refactored the code and possibly fixed some other issues with `CRateCheckBuffer`. Here it is: https://github.com/dashpay/dash/pull/1490/commits/aaeed1c46ecf127a4afe26de9b5589dce49c23c5